### PR TITLE
Implement signature help - closes #1841

### LINKED ISSUE #1841: LSP: Signature Help
I've been investigating on how to implement the `SignatureHelp` API. Can you give me a few pointers? It's pretty similar to `Hover`. The main challenge is that `Codebase::getReferenceAtPosition` can't be used because it only records the position of the identifier (excluding the parentheses). Furthermore we need to not only know which method is being called but also at what parameter the cursor is at.

### PR DESCRIPTION
**Pull request by iluuu1994**
- [x] Better tests

### CODE REVIEW

**Comment by iluuu1994** [comment]
@muglug I've noticed that the cache is not update to date on the first invocation of the signature help after a change of the source code. This problem goes away when I call `LanguageServer::doAnalysis` [first](https://github.com/vimeo/psalm/blob/master/src/Psalm/Internal/LanguageServer/LanguageServer.php#L125). What is the correct fix for this issue?

**Comment by muglug** [comment]
> What is the correct fix for this issue?

I'll be able to spend some time tomorrow on Language Server issues - will find out then!

**Comment by muglug** [comment]
You'll probably want to merge/rebase master - I removed the alias maps, as there was a better way to achieve the same thing.

**Comment by muglug** [comment]
Update: I did the rebase myself

**Comment by muglug** [comment]
> This problem goes away when I call LanguageServer::doAnalysis first

I think that may be the way forward.

I just added an explicit signature number offset which was required to get it working in Sublime Text.

**Comment by iluuu1994** [comment]
What's the reason that the `StatementsAnalyzer` stops analyzing statements when one is incorrect?

https://github.com/vimeo/psalm/blob/220fa31404ccc46a934a5a328388b3d7f19bc383/src/Psalm/Internal/Analyzer/StatementsAnalyzer.php#L574-L576

This causes the argument positions not to be recorded when a previous method call is incorrect.

**Comment by muglug** [comment]
> What's the reason that the `StatementsAnalyzer` stops analyzing statements when one is incorrect?

If an expression is doing something _really_ naughty then it might be ok to stop analysis (because the messages on the code below can be confusing).

What's the particular circumstance here? I've been removing a lot of `return false;` downstream recently to prevent this happening wherever possible.

**Comment by iluuu1994** [comment]
My test case contains multiple calls to foo:

```php
$this->foo();
$this->foo('Foo');
```

The second one won't work.

**Comment by muglug** [comment]
> The second one won't work.

It presumably doesn't work because one of the analysed expressions is returning `false` - if you can track down what's doing that (and that use case indeed doesn't seem like a serious sin) then you can have the method return `null` instead.

**Comment by muglug** [comment]
I see some low-hanging fruit, will push to master shortly.

**Comment by muglug** [comment]
see https://github.com/vimeo/psalm/commit/0e8fa0e73e3e4530ed88d457194689eabec9b905

**Comment by iluuu1994** [comment]
That solves it, thanks @muglug!

**Comment by iluuu1994** [comment]
> Why not for just the signaturehelper request?

Oh of course, that works too.

File: src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php [hunk 1]
```diff
@@ -236,6 +236,13 @@ public static function analyze(
                 strtolower($function_id)
             );
 
+            ArgumentMapPopulator::recordArgumentPositions(
+                $statements_analyzer,
+                $stmt,
```
Comment (psalm-shepherd) [inline src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallAnalyzer.php]: Argument 2 of Psalm\Internal\Analyzer\Statements\Expression\Call\ArgumentMapPopulator::recordargumentpositions expects PhpParser\Node\Expr\MethodCall|PhpParser\Node\Expr\StaticCall, PhpParser\Node\Expr\FuncCall provided
```
                $stmt,
                ^^^^^
```

**Comment by psalm-shepherd** [comment]
Psalm found some errors

**Comment by psalm-shepherd** [comment]
Psalm didn’t find any errors!

**Comment by iluuu1994** [comment]
This now works for:

* Method
* Static methods
* Functions
* Instantiations

I've encountered some super weird issue where there were `StaticCall` nodes in the AST for `class Foo extends Bar`. The `startFilePos` and `endFilePos` positions of the arguments are off completely. I have no clue what's going on there. Any ideas?

Apart from that issue this should be done.

**Comment by muglug** [comment]
Can you reproduce the weird behaviour in a test?

**Comment by iluuu1994** [comment]
I'll try tomorrow. 🙂 It happens about a dozen times when running Psalm over its own codebase.

**Comment by iluuu1994** [comment]
[This here](https://github.com/vimeo/psalm/pull/1862/files#diff-63bf0c25596421c9e72bd4c70a413599R39) is what I'm talking about. When you comment that piece of code you'll see that there are various tokenizing errors (because as mentioned the position attributes are off).

**Comment by muglug** [comment]
OHH I know what's happening - I create some AST nodes when checking for property initialisation, and those nodes are then analysed as part of that cycle. Should be a simple fix.

**Comment by iluuu1994** [comment]
Ok, I suspected something like that 😃

**Comment by muglug** [comment]
It should behave better now

**Comment by muglug** [comment]
This is fantastic - let me know if there's anything else you need from me

**Comment by iluuu1994** [comment]
Great! I've noticed that the server crashes when a signature help is invoked on native functions (`array_map` for example). I'll fix that, then it should be ready.

**Comment by iluuu1994** [comment]
@muglug Can you tell me how to get the `StatementsAnalyzer` [here](https://github.com/vimeo/psalm/pull/1862/commits/3902bb75a411a1c9e6606737f6da07a671161d17#diff-e264d1221b8ed04b20093a91a2f7650cR297)? That's the last missing piece.

**Comment by muglug** [comment]
> @muglug Can you tell me how to get the `StatementsAnalyzer` [here](https://github.com/vimeo/psalm/pull/1862/commits/3902bb75a411a1c9e6606737f6da07a671161d17#diff-e264d1221b8ed04b20093a91a2f7650cR297)?

I've just updated the function signature being called in the master branch - you'll want to merge/rebase that. But still wrap in a try/catch in case the function can't easily be discovered (I think this workaround won't work if a function is declared inside a class method).

**Comment by iluuu1994** [comment]
Done, seems to work now! 👌

**Comment by iluuu1994** [comment]
It would be nice if the popup was triggered on autocomplete, not sure if that is possible.

**Comment by muglug** [comment]
> It would be nice if the popup was triggered on autocomplete, not sure if that is possible.

I think that's up to the LSP integrator on the editor side.

**Comment by muglug** [comment]
Is this PR ready for review yet?

**Comment by iluuu1994** [comment]
> Is this PR ready for review yet?

Oh yeah sorry.

**Comment by iluuu1994** [comment]
> I think that's up to the LSP integrator on the editor side.

[Intelephense](http://intelephense.net/) can do it so it must be possible.

**Comment by muglug** [comment]
Thanks, this is great!

> It would be nice if the popup was triggered on autocomplete, not sure if that is possible.

Do you mind showing me what that looks like (via a screen recording or similar)

**Comment by iluuu1994** [comment]
This is what I'd like:

Jul-01-2019 22-05-59

This happens with Psalm:

Jul-01-2019 22-10-12

**Comment by iluuu1994** [comment]
Thanks for your help @muglug!

**Comment by muglug** [comment]
> It would be nice if the popup was triggered on autocomplete

That actually happens for me in Sublime Text, trying it out locally (one minor problem is that `<` characters are escaped properly)

file_analyzer

**Comment by muglug** [comment]
But I can reproduce the bug in VS Code

**Comment by iluuu1994** [comment]
Interesting. There must be a difference in how Intelephense does autocompletion and how Psalm does it. I'll see if I can find out.

**Comment by muglug** [comment]
I couldn’t see anything obvious in the JS source, but shouldn’t be too difficult to see the back-and-forth

**Comment by iluuu1994** [comment]
```log
[Trace - 11:06:55 PM] Received response 'textDocument/completion - (47)' in 1ms.
Result: {
    "isIncomplete": false,
    "items": [
        {
            "kind": 2,
            "label": "functionExists",
            "detail": "functionExists(StatementsAnalyzer $statements_analyzer, string $function_id)",
            "sortText": "functionExists",
            "textEdit": {
                "range": {
                    "start": {
                        "line": 51,
                        "character": 15
                    },
                    "end": {
                        "line": 51,
                        "character": 28
                    }
                },
                "newText": "functionExists($0)"
            },
            "insertTextFormat": 2,
            "command": {
                "title": "Trigger Parameter Hints",
                "command": "editor.action.triggerParameterHints"
            },
            "data": 3833510146090248
        }
    ]
}
```

Looks like it sends a `editor.action.triggerParameterHints` command. Should be super simple to implement.

**Comment by muglug** [comment]
the unescaped characters in Sublime was fixed in https://github.com/tomv564/LSP/pull/608

**Comment by muglug** [comment]
> Looks like it sends a editor.action.triggerParameterHints command. Should be super simple to implement.

Wonderful - want to do that or should I?

**Comment by iluuu1994** [comment]
Feel free, I won't have time until tomorrow.

**Comment by muglug** [comment]
Boom!

vscode_suggest

**Pull request merged**